2d_classification/MIL/mil_classify_monai.py

```diff
@@ -0,0 +1,944 @@
+
+import os
+import json
+import time
+import shutil
+import argparse
+
+import numpy as np
+from numpy.lib.stride_tricks import as_strided
+from sklearn.metrics import cohen_kappa_score
+
+import torch
+import torch.nn as nn
+# import torch.nn.functional as F
+
+from torch.utils.tensorboard import SummaryWriter
+from torch.cuda.amp import GradScaler, autocast
+from torch.utils.data.distributed import DistributedSampler
+
+import torch.nn.parallel
+import torch.distributed as dist
+import torch.multiprocessing as mp
+
+import tifffile #pip install it
+from monai import transforms, data
+
+
+def parse_args():
+
+    parser = argparse.ArgumentParser(description='MIL example')
+    parser.add_argument('--checkpoint', default=None)
+    parser.add_argument('--logdir', default=None)
+    parser.add_argument('--epochs', default=50, type=int)
+    parser.add_argument('--batch_size', default=1, type=int)
+    parser.add_argument('--fold', default=0, type=int)
+    # parser.add_argument('--checkpoint_fromscratch', action='store_true')
+
+    parser.add_argument('--optim_lr', default=3e-5, type=float)
+    parser.add_argument('--optim_trans_lr', default=None, type=float)
+
+    parser.add_argument('--quick', action='store_true') #distributed multi gpu
+    parser.add_argument('--amp', action='store_true') #experimental
+    parser.add_argument('--val_every', default=1, type=int)
+
+    parser.add_argument('--mil_mode', default='att')
+    parser.add_argument('--reg_weight', default=0, type=float)
+    parser.add_argument('--optim_trans_reg_weight', default=0, type=float)
+    parser.add_argument('--validation_only', action='store_true')
+
+    ###for multigpu
+    parser.add_argument('--distributed', action='store_true') #distributed multi gpu
+    parser.add_argument('--world_size', default=1, type=int, help='number of nodes for distributed training')
+    parser.add_argument('--rank', default=0, type=int, help='node rank for distributed training')
+    parser.add_argument('--dist-url', default='tcp://127.0.0.1:23456', type=str,  help='url used to set up distributed training')
+    parser.add_argument('--dist-backend', default='nccl', type=str, help='distributed backend')
+    parser.add_argument('--workers', default=2, type=int)
+
+    args = parser.parse_args()
+
+    print("Argument values:")
+    for k, v in vars(args).items():
+        print(k, '=>', v)
+    print('-----------------')
+
+    return args
+
+
+class AverageMeter(object):
+    """Computes and stores the average and current value"""
+    def __init__(self):
+        self.reset()
+
+    def reset(self):
+        self.val = 0
+        self.avg = 0
+        self.sum = 0
+        self.count = 0
+
+    def update(self, val, n=1):
+        if n > 0:
+            self.val = val
+            self.sum += val * n
+
+            self.count += n
+            self.avg = self.sum / self.count if self.count > 0 else self.sum
+
+
+def distributed_all_gather(tensor_list, out_numpy=False):
```

Comment:
Hi @myron ,

For this dist function, MONAI has some similar API:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/utils/dist.py#L42

Thanks.

Reply:
I looked at that function, it's similar but does a log of unnecessary calculations. it checks for the size of the first dimension of tensors, which does some padding and cropping. I just need a simple gather on tensors,  we don't need that overhead

Reply:
Hi @myron ,

When you `all_gather` PyTorch Tensors in distributed program, the Tensors may have different batch size, in that case, we must pad to make sure the first dim is same, otherwise, NCCL will hang for timeout. That's why PyTorch `DistributedSampler` drops / pads data to make data evenly divisible:
https://github.com/pytorch/pytorch/blob/master/torch/utils/data/distributed.py#L37

I think you can use the MONAI API instead, if your data shape are same, it will not pad / crop at all.

Thanks.

Reply:
my shape is the same. what is the monai API?

Reply:
Hi @myron ,

If your shape is fixed, the line with skip:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/utils/dist.py#L82
So the MONAI API is similar to your function, right?

Thanks.

Reply:
may be, that function has overhead, creating additional tensors and synchronizing on them
https://github.com/Project-MONAI/MONAI/blob/4c7a1873869a33fbc643e462d4cb167e8e7785ea/monai/utils/dist.py#L78

I want a simple and fast way, I think I can just keep this funciton in the tutorial

Reply:
OK, could you please help compare the speed of the API and your function? If your function is faster, please feel free to use it in the tutorial.
I suspect maybe they have the same speed.

Thanks.